Allows multi-type generics (ie <T, R, ...>) in method parameters

thisames opened the pull request:
Related issue

#26 

What is the purpose of this pull request?
Allow multi-type generics (ie <T, R, ...>) in method parameters

What was done?
add method _parseCommaToBar in parser.js
line 215 - used _parseCommaToBar in the parameter inside the method extractParameters loop, and change split(",") to split("|").

thisames committed:
add method _parseCommaToBar in parser.js

line 215 - used _parseCommaToBar in the parameter inside the method extractParameters loop, and change split(",") to split("|").

matheusgr commented on parser.js:
```diff
@@ -184,9 +184,35 @@ class Extractor {
 
+    _parseCommaToBar(params) {
+        let refactorString = '';
+        let foundArrow = false;
```
Use a int with `arrowDeep = 0` instead of a boolean.

At each `<` increment `arrowDeep`, and at each `>` decrement it.

This will allow multiple generics type (test case: `test: int, crazyMap: HashMap<List<Integer>, HashMap<String, List<List<Double>>>> , test: str`).

Also, nice work!

thisames replied:
Oh nice idea! I will do this.

thisames committed:
changed strategy to get generic types

add method _isGenericType in parser.js

introduced the _isGenericType method inside the _parseCommaToBar method

thisames reviewed (commented):
hmm, that test case caught me, I had to make some changes, I believe it is now working for any test case.

now basically whenever it finds a first '<' it captures all type ex: `<List<Integer>,` `HashMap<List<Integer>, HashMap<String, List<List<Double>>>> `, `<String , Integer>` checks if it is generic type and goes back to iterate in position after generic/normal type

if you find something strange, feel free to comment on it.

thisames committed:
fixed bugs.

matheusgr commented on parser.js:
```diff
@@ -184,9 +184,77 @@ class Extractor {
+        let refactorString = ''
+        let leftArrow = 0;
+
+        for (let i = 0; i < params.length; i++) {
+            if (!leftArrow) {
+                if (params[i] === ',') {
+                    refactorString += '|'
+                } else {
+                    if (params[i] !== '<') {
+                        refactorString += params[i]
+                    }
+                }
+            }
+            if (params[i] === '<') {
+                let objectGenericType = this._isGenericType(params, i)
+                if (objectGenericType.isGenericType) {
+                    refactorString += objectGenericType.stringGenericType
+                    i = objectGenericType.index + 1
+                }
+                leftArrow = 1
+            }
+            if (leftArrow) {
+                if (params[i] === ',') {
+                    refactorString += '|'
+                } else {
+                    if (params[i] !== undefined) {
+                        refactorString += params[i]
+                    }
+                }
+            }
```
you can simplify this code:

```
        let refactorString = ''
        for (let i = 0; i < params.length; i++) {
            if (params[i] === '<') {
                let objectGenericType = this._isGenericType(params, i)
                if (objectGenericType.isGenericType) {
                    refactorString += objectGenericType.stringGenericType
                    i = objectGenericType.index + 1
                }
            }
            if (params[i] === ',') {
                refactorString += '|'
            } else {
                if (params[i] !== '<' && params[i] !== undefined) {
                    refactorString += params[i]
                }
            }
        }
        return refactorString
```

Also, we need to provide a way to show an error if the generics is malformed. Example:

- `localizaUsuarios(nome: str, teste2: Map<str, ArrayList<Str>): List<String>` - missing ending '>'
- `localizaUsuarios(nome: str, teste2: Map<str, ArrayList>Str>>): List<String>` - unbalanced or strange char (">")

thisames replied:
I don't know, without checkFirstArrow, it is not possible to have the exact precision of where to place and when to place or start the iterator with '<'. without it, some tests don't pass 

 ![](https://i.imgur.com/rgAyGuO.png) 

what about errors if the generics are badly formatted, don't you want to create another issue for that? I can do it right away.

matheusgr commented on parser.js:
```diff
@@ -184,9 +184,77 @@ class Extractor {
+    }
+
     extractParameters(params) {
+        console.log(this._parseCommaToBar(params))
```
Remove all console.log before PR.

thisames replied:
ops

thisames committed:
removed console.log

thisames reviewed (commented):
removed console.log

thisames committed:
changing the logic of the _isGenericType method - greater precision in capturing types
created _semanticAnalysisGenericType method

thisames reviewed (commented):
changing the logic of the _isGenericType method - greater precision in capturing types.
created _semanticAnalysisGenericType method to analyze possible syntax errors of generic types

now it accurately captures generic types, and parses syntax errors in isolation.
